### ISSUE TITLE
no auth is not working for frontend endpoint

### ISSUE THREAD

**Issue by DimaGolomozy** [Unleash/unleash#3005]
### Describe the bug

when using locally the `unleash-server:v4.20.0` with the environment `AUTH_TYPE:NONE`
the `/api/client/...` endpoints work fine with no auth needed, this allows us to run locally unleash and test stuff without the need to provide a valid API token
but the `/api/frontend` for the client SDK still needs to pass auth header with a valid frontend API key.

this makes using unleash for local mode client testing very inconvenient, as each developer will need to spin up a local `unleash` server, login, and create a frontend token. we wanted to reduce those steps by providing `AUTH_TYPE:NONE`

### Steps to reproduce the bug

1. sping unleash up with docker compose:
```
    unleash-db:
        container_name: unleash-db
        image: postgres:13.5-alpine3.14
        environment:
            POSTGRES_PASSWORD: "unleash"
            POSTGRES_USER: "unleash"
            POSTGRES_DB: "unleash"
            PGUSER: "unleash"
        healthcheck:
            test: [ "CMD", "pg_isready", "--username=unleash", "--host=127.0.0.1", "--port=5432" ]
            interval: 2s
            timeout: 1m
            retries: 5
            start_period: 5s

    unleash:
        container_name: unleash
        image: unleashorg/unleash-server:4.20.0
        ports:
            - "4242:4242"
        environment:
            DATABASE_HOST: "unleash-db"
            DATABASE_NAME: "unleash"
            DATABASE_USERNAME: "unleash"
            DATABASE_PASSWORD: "unleash"
            DATABASE_SSL: "false"
            LOG_LEVEL: "debug"
            ENABLED_ENVIRONMENTS: "default"
            AUTH_TYPE: "NONE"
        healthcheck:
            test: [ "CMD", "nc",  "-z", "unleash-db", "5432" ]
            interval: 1s
            timeout: 1m
            retries: 5
            start_period: 10s
        depends_on:
            - unleash-db
```

2. simple curl that works fine:
```
curl http://localhost:4242/api/client/features
```
3. curl the frontend endpoint to see the error:
```
curl 'http://localhost:4242/api/frontend?appName=dima&environment=default'
```

this endpoint is not working and also there are errors on the unleash server:
```
[2023-01-27T06:47:47.709] [WARN] controller/ProxyController - Cannot destructure property 'type' of 'undefined' as it is undefined.
[2023-01-27T06:47:47.709] [ERROR] controller/ProxyController - Server failed executing request TypeError: Cannot destructure property 'type' of 'undefined' as it is undefined.
    at Function.assertExpectedTokenType (/unleash/node_modules/unleash-server/dist/lib/services/proxy-service.js:78:38)
    at ProxyService.clientForProxyToken (/unleash/node_modules/unleash-server/dist/lib/services/proxy-service.js:52:22)
    at ProxyService.getProxyFeatures (/unleash/node_modules/unleash-server/dist/lib/services/proxy-service.js:25:35)
    at ProxyController.getProxyFeatures (/unleash/node_modules/unleash-server/dist/lib/routes/proxy-api/index.js:99:56)
    at /unleash/node_modules/unleash-server/dist/lib/routes/controller.js:37:23
    at Layer.handle [as handle_request] (/unleash/node_modules/express/lib/router/layer.js:95:5)
    at next (/unleash/node_modules/express/lib/router/route.js:144:13)
    at validateMiddleware (/unleash/node_modules/@unleash/express-openapi/lib/validate.js:105:14)
    at validSchemaMiddleware (/unleash/node_modules/@unleash/express-openapi/index.js:71:14)
    at Layer.handle [as handle_request] (/unleash/node_modules/express/lib/router/layer.js:95:5)
    at next (/unleash/node_modules/express/lib/router/route.js:144:13)
    at /unleash/node_modules/unleash-server/dist/lib/routes/controller.js:13:16
    at Layer.handle [as handle_request] (/unleash/node_modules/express/lib/router/layer.js:95:5)
    at next (/unleash/node_modules/express/lib/router/route.js:144:13)
    at Route.dispatch (/unleash/node_modules/express/lib/router/route.js:114:3)
    at Layer.handle [as handle_request] (/unleash/node_modules/express/lib/router/layer.js:95:5) {
  isJoi: true
}
```

### Expected behavior

when making `AUTH_TYPE:OPEN_SOURCE` and making the curl command (with the frontend auth token) it works fine and returns the `toggles`
the expected behavior is when using `AUTH_TYPE:NONE` the frontend endpoint should not require a token 

### Subscription type

Open source

### Hosting type

Self-hosted

**Comment by sighphyre** [comment]
@DimaGolomozy Thanks for raising this. Looks to be an oversight on our part. We'll dig a little deeper next week and update this when we have progress

**Comment by sighphyre** [comment]
Hey @DimaGolomozy, so I've looked into this one and I'm not actually sure the `api/client` endpoints are working as you expect. The no auth option is intended to make the UI easy to use but not necessarily the endpoints that the SDKs use - we use the incoming token to determine the permissions and environments that should be considered when returning the response. From reading the code and testing a little, when you use no auth and you hit the `api/client` endpoints you will get a 200 response but it won't contain what you want. Would you mind sanity checking me on your cURL call - you should get a list of features but no strategies and the toggles will be disabled.

Typically how I would recommend making this easy to use is to spin up Unleash and generate the API keys you need with cURL. You can then use those for local testing. Same approach should work for frontend keys

**Comment by DimaGolomozy** [comment]
settings `AUTH_TYPE: "NONE"`
this is a result of postman with no auth header provided. you can see the features are returned. (one feature with `"enabled" :true" and strategies)
<img width="1439" alt="image" src="https://user-images.githubusercontent.com/9161830/216035713-7944e742-9d91-4f8f-b7bc-b6dfb8b86bb4.png">

and this is what i get when trying the `frontend` endpoint.  
(see the `500 Internal Server Error`, the log is provided in the issue description)
<img width="1368" alt="image" src="https://user-images.githubusercontent.com/9161830/216036204-60b4f469-8f61-4299-9c63-32dd12920057.png">

what I expect, is for the `frontend` endpoint to act the same as the `api/client` endpoint.

**Comment by thomasheartman** [comment]
This was moved to done in our issue board 10 months ago and nothing has happened here since. I'm going to close this issue now. @sighphyre if you think it should remain open, feel free to chime in when you're back.

**Comment by david-annopolsky** [comment]
Hi @thomasheartman what is the solution im running into the same issue as well

**Comment by thomasheartman** [comment]
Hey, @david-annopolsky. I don't think there is a solution. As @sighphyre mentioned in his response over there, the client APIs rely on the auth token to be able to read access permissions and environment data. With no token, you'd probably only get flags for the "default" environment, which is deprecated and only available to users who have been around since the 4.x days.

From a discussion in a linked PR (#3031):

> I've consulted with @sighphyre on this, and we agree that AuthType.\<whatever\> should determine the login flow to the admin UI and **not affect the client API**. Because the client API is dependent on tokens to provide important information such as environment and projects, the tokens will be required for that regardless of AuthType. 

So in short, I think this is working as expected.

**Comment by david-annopolsky** [comment]
Hi @thomasheartman thanks for the quick response, we are using none for auth-type in admin ui, but then passing in a token to our `/api/frontend` even generated from our environment admin ui results in 

>  Cannot destructure property 'type' of 'undefined' as it is undefined. 

Looking at the code for the error stack it looks like the token type which should be either frontend or admin isn't being passed 

https://github.com/Unleash/unleash/blob/main/src/lib/services/proxy-service.ts#L157

any thoughts or ideas with that?

we are running unleash-server and react app sdk on our aws env

**Comment by david-annopolsky** [comment]
small update removing the `auth` none from the `unleash-server` and providing the token generated in admin ui works on the `api/frontend` but if i provide type NONE for authentication i get an error about destructuring which i mentioned in the above comment.

```js
sync function start() {
  try {
    const unleashed = await unleash.start({
      //authentication: {
       // type: IAuthType.NONE,
      //},
   
```

@thomasheartman @sighphyre any thoughts on this?

**Comment by thomasheartman** [comment]
@david-annopolsky  thanks for reporting back! And yes, I think you're right. I've been thinking you might hitting the same issue as what #3031 was supposed to solve. Looking back over that PR, I think we should take the time to add the tests for it and close it ourselves. That might address your issue. 

I'm a little off-grid at the moment, though, so not able to take a super detailed look at this until hopefully tomorrow. Shoot me a message if you don't hear anything back from me by end of day tomorrow 🙋🏼

**Comment by thomasheartman** [comment]
@david-annopolsky Thanks for clarifying. To make sure we're on the same page, is this correct?
You'd like to:
- Use the admin UI without auth
- Use API tokens for connecting applications to Unleash's client APIs

Is that correct?

**Comment by thomasheartman** [comment]
@david-annopolsky Just wanted to report that we've reproduced the issue and that we're working on writing the necessary tests for the code before pushing this out. Hopefully shouldn't be long. Thanks for pushing it 🙏🏼

**Comment by david-annopolsky** [comment]
> Just wanted to report that we've reproduced the issue and that we're working on writing the necessary tests for the code before pushing this out. Hopefully shouldn't be long. Thanks for pushing it

@thomasheartman  you are saying there is a fix coming into place that will allow us to have an admin ui without auth **but** connecting a client sdk to unleash-server via API tokens?

and what about if i want to turn off `enableApiToken`?

**Comment by thomasheartman** [comment]
@david-annopolsky 

> you are saying there is a fix coming into place that will allow us to have an admin ui without auth but connecting a client sdk to unleash-server via API tokens?

Yeah, that is what I'm saying. However, it turns out to be a little more difficult than expected and we need to figure some things out. I thought it would just be writing a test for the PR mentioned earlier and merging that, but it turns out that we rely on some of that behavior in our own tests. So we need to decide what the best way forward here is.

> and what about if i want to turn off enableApiToken?

That's a good question. I have to admit that I was not aware of that option (until you mentioned it a few comments ago). I suspect that this is a holdover from an earlier age, when Unleash had fewer bells and whistles. The way it is today, it doesn't make sense to access Unleash client APIs without a token because you'll only get data for the default environment, which isn't available anymore except for in legacy installations. 

But I've raised this internally. Due to time zone differences, it'll be a little while before I hear something, but I'll try to update you once I hear something.

**Comment by thomasheartman** [comment]
So after discussing this internally, here's what we have come to:
We'll work on fixing the issue where using API tokens for the front-end API with auth type none doesn't work. We might also make it work without tokens to reflect how the client/features API works. Note, though, that you probably don't want to use it this way because you can't set the environment access.

As for what `enableApiTokens` does and should do: this is more uncertain. It was an option introduced a long time ago that hasn't been maintained. How it works today is largely by luck.

Did you have a use case where you needed to use the front-end API without an API token or was the question mostly out of curiosity?

**Comment by david-annopolsky** [comment]
> Did you have a use case where you needed to use the front-end API without an API token or was the question mostly out of curiosity?

yes in our case since we are currently self hosting the application, i didnt see a path forward where we have API tokens can be create automatically and provided to the frontend without manual intervention. meaning i think i saw a way to start unleash-server via `initApiTokens ` (i believe thats the idea for that) but i dont have a way of having the frontend app to have that same token without committing in code and then that means each environment we share the same hardcode token (seems not ideal?) or we can hardcode different tokens for each environment but again that doesnt seems ideal.

All of our client side apps are very simple and have servers they connect to via auth header token from our auth provider so there is no access to the client side apps without logging in first and that means no unauthed access to server side which would include the unleash-server so ideally we wouldnt want to connect to unleash via another API token because unleash-server technically could support  custom hook (plugin/middleware like admin ui auth) that could just validate the incoming token in the header or something like that (but thats just something that came to my mind just now)

let me know your thoughts please

**Comment by thomasheartman** [comment]
> yes in our case since we are currently self hosting the application, i didnt see a path forward where we have API tokens can be create automatically and provided to the frontend without manual intervention. meaning i think i saw a way to start unleash-server via initApiTokens (i believe thats the idea for that) but i dont have a way of having the frontend app to have that same token without committing in code and then that means each environment we share the same hardcode token (seems not ideal?) or we can hardcode different tokens for each environment but again that doesnt seems ideal.

You're right. You can start unleash with `initApiTokens` (or the env var `INIT_FRONTEND_API_TOKENS`) to start Unleash with a set of tokens. How you get that to your front end applications would be up to you. Indeed, committing the token isn't ideal, but for client-side apps, it's not really an issue. Anything that is in users' hands they can inspect anyway, and the front end API is designed such that it does not leak any information outside of just what is enabled for the current user. And even if you didn't commit the token, the token would still be visible to the user when the request is made anyway.

Besides, there's very little functional difference between having no auth and a static token that is visible to the user. The biggest difference I can see is that with the static token, you can at least restrict access down to environment and projects, whereas without auth, you would have no way of doing that.

Now, practically, it's a little more work: you'd actually need to put the token in either in the source code or in some other place where it gets picked up when deploying (you could probably also inject it when you serve the client code? injected from the server somehow?), but it doesn't seem like a lot. Then again, I don't know your full setup, so it might be more work than I imagine.

> All of our client side apps are very simple and have servers they connect to via auth header token from our auth provider so there is no access to the client side apps without logging in first and that means no unauthed access to server side which would include the unleash-server so ideally we wouldnt want to connect to unleash via another API token because unleash-server technically could support custom hook (plugin/middleware like admin ui auth) that could just validate the incoming token in the header or something like that (but thats just something that came to my mind just now)

Right. I think I might see what you're saying. If that's how you want to do it, you could probably build a super-simple proxy that takes your auth header and just adds the correct auth token before passing the request on to Unleash? That way, you wouldn't need the client-side apps to be aware of the token; they'd just use whatever auth you're using for the rest of your system. And then you use this little transformer to add the right token based on whatever info you already have in the request (to figure out environment, projects, or whatever else you need), before passing it on Unleash. Could that work? (Of course, it requires that we fix the frontend auth bug; I'm expecting to know more about that later today).

But it does really come down to this: Unleash requires the token to know what environment you want flags for (project access is optional, but environment is not). There is just no way to get around this, I'm afraid, and I don't think we're planning on making it possible in the future either.

**Comment by david-annopolsky** [comment]
@thomasheartman thanks for the detailed write up and understood about the tokens, i see an issue has been created for this bug, do you know how long it may take for a resolution

**Comment by thomasheartman** [comment]
Of course; happy to help 😄 As for the time to resolution, I'm not sure. It's on @daveleek's plate right now, so he'd be the one to ask. If it's a quick fix and nothing more urgent pops up in the meantime, then it should be quick. But I'm not sure how much work it would require or what Dave's workload is like today. Of course, you're free to try to implement it yourself (either as a temporary local fix or to open a PR for it), but let's hear what Dave thinks.

**Comment by thomasheartman** [comment]
Little update: @daveleek has opened a #5806, which should fix this when it's ready, so you can check in with that issue to know what's up 😄

**Comment by david-annopolsky** [comment]
thanks for the updates, we are currently trying out a custom handler with openid connect as a workaround thanks again for the support and the conversation
